nomad/eval_endpoint.go

```diff
@@ -416,6 +417,149 @@ func (e *Eval) Reap(args *structs.EvalDeleteRequest,
+// differs from Reap while duplicating some behavior to ensure we have the
+// correct controls for user initiated deletions.
+func (e *Eval) Delete(
+	args *structs.EvalDeleteRequest,
+	reply *structs.EvalDeleteResponse) error {
+
+	if done, err := e.srv.forward(structs.EvalDeleteRPCMethod, args, args, reply); done {
+		return err
+	}
+	defer metrics.MeasureSince([]string{"nomad", "eval", "delete"}, time.Now())
+
+	// This RPC endpoint is very destructive and alters Nomad's core state,
+	// meaning only those with management tokens can call it.
+	if aclObj, err := e.srv.ResolveToken(args.AuthToken); err != nil {
+		return err
+	} else if aclObj != nil && !aclObj.IsManagement() {
+		return structs.ErrPermissionDenied
+	}
+
+	// The eval broker must be disabled otherwise Nomad's state will likely get
+	// wild in a very un-fun way.
+	if e.srv.evalBroker.Enabled() {
+		return errors.New("eval broker is enabled")
+	}
+
+	// Grab the state snapshot, so we can look up relevant eval information.
+	serverStateSnapshot, err := e.srv.State().Snapshot()
+	if err != nil {
+		return fmt.Errorf("failed to lookup state snapshot: %v", err)
+	}
+	ws := memdb.NewWatchSet()
+
+	// Iterate the evaluations and ensure they are safe to delete. It is
+	// possible passed evals are not safe to delete and would make Nomads state
+	// a little wonky. The nature of the RPC return error, means a single
+	// unsafe eval ID fails the whole call.
+	for _, evalID := range args.EvalIDs {
+
+		evalInfo, err := serverStateSnapshot.EvalByID(ws, evalID)
+		if err != nil {
+			return fmt.Errorf("failed to lookup eval: %v", err)
+		}
+		if evalInfo == nil {
+			return errors.New("eval not found")
+		}
+
+		jobInfo, err := serverStateSnapshot.JobByID(ws, evalInfo.Namespace, evalInfo.JobID)
+		if err != nil {
+			return fmt.Errorf("failed to lookup eval job: %v", err)
+		}
+
+		allocs, err := serverStateSnapshot.AllocsByEval(ws, evalInfo.ID)
+		if err != nil {
+			return fmt.Errorf("failed to lookup eval allocs: %v", err)
+		}
+
+		if !evalDeleteSafe(allocs, jobInfo) {
+			return fmt.Errorf("eval %s is not safe to delete", evalID)
+		}
+	}
+
+	// Generate the Raft request object using the reap request object. This
+	// avoids adding new Raft messages types and follows the existing reap
+	// flow.
+	raftReq := structs.EvalReapRequest{
+		Evals:        args.EvalIDs,
+		WriteRequest: args.WriteRequest,
+	}
+
+	// Update via Raft.
+	_, index, err := e.srv.raftApply(structs.EvalDeleteRequestType, &raftReq)
+	if err != nil {
+		return err
+	}
+
+	// Update the index and return.
+	reply.Index = index
+	return nil
+}
+
+// evalDeleteSafe ensures an evaluation is safe to delete based on its related
+// allocation and job information. This follows similar, but different rules to
+// the eval reap checking, to ensure evaluations for running allocs or allocs
+// which need the evaluation detail are not deleted.
+func evalDeleteSafe(allocs []*structs.Allocation, job *structs.Job) bool {
```

Comment:
No, but we only allow deleting when the eval broker is paused and the queue is re-populated after being unpaused.

However, you made me realize we don't enforce the eval broker is paused in the *fsm* which is required to ensure deletes don't interleave with queue pauses/unpauses.

In the Periodic GC case we're only deleting terminal evals which cannot be in the eval broker(s) anyway, so there's no need to worry about mutating the state store in a way that makes it inconsistent with the queues.

However in this delete path I don't think there's anything stopping someone from unpausing the queue during an eval delete.

Luckily the queue is paused/resumed from Raft so we can serialize these operations! I think the FSM handler for deleting evals needs to check `SchedulerConfig.PauseBroker` whenever being told to delete an eval manually and refuse to if the broker is running.

Reply:
A wrench in this: the scheduler config isn't in raft, right? So if we check it in the FSM we could have inconsistent views between the leader and the followers when we apply the entry and then the followers wouldn't apply the replicated logs.

Reply:
Good points.

I think in order to achieve this without the need for a new Raft message I can add a new boolean field on the request that gets sent via Raft which is used to indicate whether the operation was user initiated. This will default to false mimicking the reap behaviour, and the new RPC eval delete endpoint will set this to true which will then force the state store to perform a lookup within a single transaction to confirm the broker is paused.